src/ModelContextProtocol/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -23,23 +20,17 @@ public static partial class McpServerBuilderExtensions
     /// <typeparam name="TTool">The tool type.</typeparam>
     /// <param name="builder">The builder instance.</param>
     /// <exception cref="ArgumentNullException"><paramref name="builder"/> is <see langword="null"/>.</exception>
-    public static IMcpServerBuilder WithTool<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] TTool>(this IMcpServerBuilder builder)
+    public static IMcpServerBuilder WithTools<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)] TTool>(
+        this IMcpServerBuilder builder)
     {
         Throw.IfNull(builder);
-        List<AIFunction> functions = [];
 
-        PopulateFunctions(typeof(TTool), functions);
-        return WithTools(builder, functions);
-    }
-    /// <summary>
-    /// Adds all tools marked with <see cref="McpToolTypeAttribute"/> from the current assembly to the server.
-    /// </summary>
-    /// <param name="builder">The builder instance.</param>
-    /// <exception cref="ArgumentNullException"><paramref name="builder"/> is <see langword="null"/>.</exception>
-    [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
-    public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder)
-    {
-        return WithToolsFromAssembly(builder, Assembly.GetCallingAssembly());
+        foreach (var toolMethod in GetToolMethods(typeof(TTool)))
+        {
+            builder.Services.AddSingleton(services => McpServerTool.Create(toolMethod, services: services));
```

Comment:
I'm trying to understand what the meaning of chaining multiple `AddSingleton` calls on the same type is. Is it not the case that the previous entries are being overwritten?

Reply:
> Is it not the case that the previous entries are being overwritten?

No. They all end up in DI, and they're all available via GetService, which returns an enumerable of them all. If you ask for just one, you'll get just one, of course.